### PR TITLE
18 Restructure and Testing

### PR DESCRIPTION
**Pull request by secretsparkle**
Hey guys, this branch is largely focused on a file restructure and testing. For the review, focus mainly on the testing in special_forms_test.go and math_test.go. The rest of the tests have placeholder code in them.

### CODE REVIEW

**Comment by secretsparkle** [comment]
And Emma, you are familiar with the old file structure, any comments on the new file structure?

File: engines/io.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	args = args[1:] //shave off that opening paren
+	expressions, _, err := parse.Transliterate(expressions, args, 0)
+
+	/*
+		if err = parse.Parse(input); err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+	*/
+	return expressions, err
+}
```
Comment (emkristo) [inline engines/io.go:24]: I can't decide if I want this to be in parse

File: evaluation/.#functions.go [hunk 2]
```diff
@@ -0,0 +1 @@
+scott@Scotts-MacBook-Pro.local.55676
```
Comment (emkristo) [inline evaluation/.#functions.go:1]: Delete this file

File: evaluation/math_test.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	if test6 != 1 {
+		t.Errorf("(+ 1) = %v; want 1", test6)
+	}
+	if err7 == nil {
+		t.Errorf("(+) = %v; want invalid number of arguments.", err7)
+	}
+	if err8 == nil {
+		t.Errorf("(+ $ #) = %v; want invalid syntax.", err8)
+	}
+}
```
Comment (emkristo) [inline evaluation/math_test.go:42]: suggested addition: (+ '(1 2 3) '(4 5 6)) should say it doesn't work on lists or something like that, same with adding bools; do the same thing for all arithmetic operations

File: evaluation/special_form_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	if test9 != false {
+		t.Errorf("(and t nil) = %v; want nil", test9)
+	}
+	if test10 != false {
+		t.Errorf("(and nil t) = %v; want nil", test10)
+	}
+	if err11 == nil {
+		t.Errorf("(and $test nil) = %v; want error", err11)
+	}
+}
```
Comment (emkristo) [inline evaluation/special_form_test.go]: Perhaps add more tests such as 

```
> (and nil)
NIL
> (and (not a function) nil)
Invalid list construction
> (and nil (not a function))
NIL
> (and nil $test)
NIL
> (and '(1 2 3) '(4 5 6))
(4 5 6)
> (and)
T
```

Note that `(and)` is a bug at the moment, or at least SBCL returns T

**Review by emkristo — COMMENTED** [review]
Excellent work, Scott! The file structure is definitely a lot more organized. 

I have a suggestion for structuring the tests. It may be easier to follow if you make a list of tests and iterate through them like the following example for TestPlus:

```
func TestPlus(t *testing.T) {
    types, functions, bindings := structs.Maps()

    
    successful_tests := map[string]int{
        "(+ 1 2)"               :   3,
        "(+ 1 2 3)"             :   6,
        "(+ 1 -4)"              :   -3,
        "(+ 1 (+ 1 2))"         :   4,
        "(+ 1 (+ (+ 1 2) 1))"   :   5,
        "(+ 1)"                 :   1
    }

    // Maybe just make this a list if it's always just nil or not
    unsuccessful_tests := map[string]interface{}{
        "(+)"                   :   nil,    // idk what should go here
        "(+ $ #)"               :   nil     // idk what should go here
    }

    for test, desired := range successful_tests {
        expressions, _ = engines.Translate(test)
        res, _ := plus(expressions, &types, &functions, &bindings)

        if res != desired {
            t.Errorf("print error message")
        }
    }

    for test, undesired := range unsuccessful_tests {
        expressions, _ = engines.Translate(test)
        _, err := plus(expressions, &types, &funtions, &bindings)

        if err == undesired {
            t.Errorf("print error message")
        }
    }    
}
```

Alternatively, you could try using the testing structure in the files I emailed you. The structure they used was some fancy golang thing, but it's the same concept.

**Comment by emkristo** [comment]
> And Emma, you are familiar with the old file structure, any comments on the new file structure?

Yes I love it and I'm your biggest fan.

File: evaluation/conditionals.go [hunk 5]
```diff
@@ -55,9 +69,18 @@ func equal(expression structs.List, symbols *map[string]rune,
...
 			b, _ = strconv.ParseFloat(b.(string), 64)
+		} else if e.Data.(string) == "t" || e.Data.(string) == "T" {
+			b = e.Data
+		} else if e.Data.(string) == "nil" || e.Data.(string) == "NIL" {
+			b = false
+		} else if util.IsAlphabetic(e.Data.(string)) {
+			return e.Data, errors.New("Unbound symbol, cannot evaluate")
+		} else if util.AnySymbol(e.Data.(string)) {
+			return e.Data, errors.New("Cannot evaluate symbolic input")
+		} else if util.IsNumber(e.Data.(string)) {
```
Comment (emkristo) [inline evaluation/conditionals.go:80]: This is repeated in lines 32-41. Consider making a function for this so you don't have to copy/paste as much. Like a type-checker or something.

File: evaluation/conditionals.go [hunk 6]
```diff
@@ -121,6 +144,10 @@ func gt_or_lt(expression structs.List, symbols *map[string]rune,
 	var err error
 	var str interface{}
 
+	if expression.Len() != 3 {
+		return nil, errors.New("Invalid number of arguments")
+	}
```
Comment (emkristo) [inline evaluation/conditionals.go:149]: Lines 14-17 are also repeated. I know it's short and not a big deal, but something like

`assert_length(expression, 3)`

would be clean. Returning is the hard part. Talk to me if you want ideas.

File: evaluation/conditionals_test.go [hunk 7]
```diff
@@ -9,32 +9,110 @@ import (
...
-	test1, _ := and(expressions, &types, &functions, &bindings)
+	successful_tests := map[string]bool{
+		"(> 3 4)":       false,
+		"(> 3 3)":       false,
+		"(> 4 3)":       true,
+		"(> (+ 3 3) 4)": true,
+		"(> 4 (+ 3 3))": false,
+	}
+
+	unsuccessful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/conditionals_test.go:59]: add test for lists failing

File: evaluation/conditionals_test.go [hunk 8]
```diff
@@ -9,32 +9,110 @@ import (
...
-	test1, _ := and(expressions, &types, &functions, &bindings)
+	successful_tests := map[string]bool{
+		"(< 3 4)":       true,
+		"(< 3 3)":       false,
+		"(< 4 3)":       false,
+		"(< (+ 3 3) 4)": false,
+		"(< 4 (+ 3 3))": true,
+	}
+
+	unsuccessful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/conditionals_test.go:95]: add fail case for list

File: evaluation/lists.go [hunk 9]
```diff
@@ -203,6 +218,9 @@ func reverse(expression structs.List, symbols *map[string]rune, functions *map[s
 	bindings *map[string]string) (interface{}, error) {
 	var reversed structs.List
 
+	if expression.Len() != 2 {
```
Comment (emkristo) [inline evaluation/lists.go:221]: See what I mean about all these if statements to check the length? This is why I love making stupid little assert functions.

File: evaluation/lists_test.go [hunk 10]
```diff
@@ -9,65 +9,203 @@ import (
...
+		}
 	}
 }
 
 func TestCons(t *testing.T) {
 	types, functions, bindings := structs.Maps()
 
-	expressions, _ := engines.Translate("(and t)")
-	test1, _ := and(expressions, &types, &functions, &bindings)
+	successful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/lists_test.go:80]: how about `(cons 1 2)`  and `(cons '(1 2) '(3 4))` and `(cons '(2 3) 1)` and `(cons nil '(T nil))`

File: evaluation/conditionals_test.go [hunk 11]
```diff
@@ -9,32 +9,110 @@ import (
 func TestEqual(t *testing.T) {
 	types, functions, bindings := structs.Maps()
 
-	expressions, _ := engines.Translate("(and t)")
-	test1, _ := and(expressions, &types, &functions, &bindings)
+	successful_tests := map[string]bool{
```
Comment (emkristo) [inline evaluation/conditionals_test.go:12]: `(equal nil '())`

Comment (emkristo) [reply evaluation/conditionals_test.go:12]: Also maybe comparing lists and numbers to bools

File: evaluation/lists_test.go [hunk 12]
```diff
@@ -9,65 +9,203 @@ import (
...
+		}
 	}
 }
 
 func TestLast(t *testing.T) {
 	types, functions, bindings := structs.Maps()
 
-	expressions, _ := engines.Translate("(and t)")
-	test1, _ := and(expressions, &types, &functions, &bindings)
+	successful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/lists_test.go:115]: `(last '(1))` and `(last '())`

File: evaluation/lists_test.go [hunk 13]
```diff
@@ -9,65 +9,203 @@ import (
...
+		_, err := last(expression, &types, &functions, &bindings)
+		if err == undesired {
+			t.Errorf("%s triggered %v; want error triggered.", test, err)
+		}
 	}
 }
 
 func TestList(t *testing.T) {
-	types, functions, bindings := structs.Maps()
+	successful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/lists_test.go:147]: how about some bools

File: evaluation/lists_test.go [hunk 14]
```diff
@@ -9,65 +9,203 @@ import (
...
+		}
 	}
 }
 
 func TestReverse(t *testing.T) {
 	types, functions, bindings := structs.Maps()
 
-	expressions, _ := engines.Translate("(and t)")
-	test1, _ := and(expressions, &types, &functions, &bindings)
+	successful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/lists_test.go:182]: `(reverse (list T (list) (list)))` and `(reverse (list))`

File: evaluation/math_test.go [hunk 15]
```diff
@@ -9,211 +9,136 @@ import (
...
+		"(+ 1 2 3)":           6,
+		"(+ 1 -4)":            -3,
+		"(+ 1 (+ 1 2))":       4,
+		"(+ 1 (+ (+ 1 2) 1))": 5,
+		"(+ 1)":               1,
+	}
 
-	expressions, _ = engines.Translate("(+ 1)")
-	test6, _ := plus(expressions, &types, &functions, &bindings)
+	unsuccessful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/math_test.go:21]: bools

File: evaluation/special_forms_test.go [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+import (
+	"../engines"
+	"../structs"
+	"testing"
+)
+
+func TestAnd(t *testing.T) {
+	types, functions, bindings := structs.Maps()
+
+	successful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/special_forms_test.go:12]: `(and 3 3)`

File: evaluation/special_forms_test.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+	if test1 != "t" {
+		t.Errorf("(and t) = %v; want t", test1)
+	}
+}
+
+func TestIfStatement(t *testing.T) {
+	types, functions, bindings := structs.Maps()
+
+	successful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/special_forms_test.go:72]: `(if (list) (notathing) '(1))`

File: evaluation/special_forms_test.go [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+		"(if (and 1 nil) 4 t)": true,
+		"(if t t)":             true,
+		"(if '(1 2 3) 7 8)":    "7",
+		"(if 2 3 4)":           "3",
+		"(if (< 2 3) 2)":       "2",
+		"(if (equal 3 3) 4 8)": "4",
+		"(if (equal 3 4) 4 8)": "8",
+	}
+
+	unsuccessful_tests := map[string]interface{}{
```
Comment (emkristo) [inline evaluation/special_forms_test.go:85]: `(if (list) (list) (notathing))`

**Review by emkristo — APPROVED** [review]
You've done so great! Lots of tests :)

Add my suggestions you feel are appropriate then you can pass through or resolve comments if you're still working on the branch

**Pull request merged**